### ISSUE TITLE
Valid disables reported when `--report-needless-disables` and `--quiet` are set

### ISSUE THREAD

**Issue by aaronccasanova** [stylelint/stylelint#7843]
### What minimal example or steps are needed to reproduce the bug?

- Add a custom plugin using `stylelint.utils.checkAgainstRule()`

```js
import stylelint from "stylelint";

const ruleName = "custom/plugin";

function rule(primary, secondary, context) {
  return async (root, result) => {
    await stylelint.utils.checkAgainstRule(
      {
        ruleName: "block-no-empty",
        ruleSettings: [primary, secondary],
        root,
        result,
        context
      },
      (warning) => {
        stylelint.utils.report({
          ruleName,
          result,
          message: warning.text,
          node: warning.node,
        });
      }
    );
  };
};

export default stylelint.createPlugin(ruleName, rule);
```

- Add an `example.css` file with the `checkAgainstRule` disabled

```css
/* stylelint-disable */
a {}
```

### What minimal configuration is needed to reproduce the bug?

```js
export default {
  plugins: ["./plugins/custom-plugin"],
  rules: { "custom/plugin": true },
};
```

### How did you run Stylelint?

```shell
stylelint example.css --report-needless-disables
```
```shell
stylelint example.css --report-needless-disables --quiet
```

### Which Stylelint-related dependencies are you using?

```json
{
  "devDependencies": {
    "stylelint": "16.7.0"
  }
}
```

### What did you expect to happen?

Expected both runs of Stylelint to report no errors

### What actually happened?

- ✅ Stylelint with `--report-needless-disables` option set reports no errors
- ❌ Stylelint with `--report-needless-disables` and `--quiet` options set reports errors

![image](https://github.com/user-attachments/assets/2f29313b-a1a0-446f-8694-a8808150cc62)

### Do you have a proposal to fix the bug?

I don't have a proposal. However, I did link `stylelint` to my project, stepped back through the history, and identified that the failures started with this [PR](https://github.com/stylelint/stylelint/pull/7833). Additionally, I began reverting changes line by line and pinpointed the issue to [this change in checkAgainstRule](https://github.com/stylelint/stylelint/pull/7833/files#diff-7d9680e3687d5b028c2a7a394026ed850a4d45b704584a2d9001c86d32006d34R46).

**Comment by Mouvedia** [comment]
> Additionally, I began reverting changes line by line and pinpointed the issue to [this change in checkAgainstRule](https://github.com/stylelint/stylelint/pull/7833/files#diff-7d9680e3687d5b028c2a7a394026ed850a4d45b704584a2d9001c86d32006d34R46).

Maybe it is due to [`StylelintPostcssResult`](https://github.com/stylelint/stylelint/blob/main/types/stylelint/index.d.ts#L156) containing `quiet` which used to be reset to `undefined`?

It makes me wonder why we have `quiet` on both `Config` and `StylelintPostcssResult`.
https://github.com/stylelint/stylelint/pull/380/files#diff-598b75d6dc772eb13aa7141f35bb93046ae26c58a97e8647ecf7f74aec8ae93fR42

**Comment by ybiquitous** [comment]
@aaronccasanova Thanks for the report with a reproduction. I believe this is a bug in `report()`. See the debug output below:

```diff
--- a/lib/utils/report.mjs
+++ b/lib/utils/report.mjs
@@ -28,6 +28,7 @@ export default function report(problem) {
 	} = result.stylelint;
 	const { messageArgs = [], severity = ruleSeverities[ruleName] } = rest;
 	const ruleSeverity = isFn(severity) ? severity(...messageArgs) || defaultSeverity : severity;
+	console.log({ quiet, defaultSeverity, severity, ruleSeverity, ruleSeverities, ruleName });
 
 	// In quiet mode, mere warnings are ignored
 	if (quiet && ruleSeverity !== 'error') return;
```

```sh-session
$ bin/stylelint.mjs -c tmp/stylelint.config.mjs tmp/test.css --report-needless-disables --quiet
{
  quiet: true,
  defaultSeverity: 'error',
  severity: undefined,
  ruleSeverity: undefined,
  ruleSeverities: { 'custom/foo': 'error' },
  ruleName: 'block-no-empty'
}

tmp/test.css
  1:1  ✖  Needless disable for "all"  --report-needless-disables
```

https://github.com/stylelint/stylelint/blob/1ba1a0aeafaa124a4987a0135d511111cec3af0a/lib/utils/report.mjs#L30-L33

It seems the `ruleSeverity` variable should not be `undefined`.

**Comment by ybiquitous** [comment]
> It makes me wonder why we have `quiet` on both `Config` and `StylelintPostcssResult`.

We may be able to refactor the duplicate `quiet`, but it seems out of the scope of this bug.

**Comment by Mouvedia** [comment]
We should have the same bug for `customMessages` and `customUrls`.
The `ruleName` used is the one passed and not the one registered in `result`.
i.e. `"custom/plugin" !== "block-no-empty"`

**Comment by ybiquitous** [comment]
Right. This is a bug. Let's fix it.

**Comment by Mouvedia** [comment]
> Let's fix it.

Just to be sure.
Would replacing the key, if found, in `customMessages`, `customUrls` and `disabledRanges` by the extending `ruleName` be the fix you are talking about?

**Comment by Mouvedia** [comment]
> Let's fix it.

Just to be more specific, after this fix the file `example.css`:

```css
/* stylelint-disable custom/plugin */
a {}
```

will not report an error for

```sh
stylelint example.css --report-needless-disables --quiet
```

and tests will have to be added in `checkAgainstRule.test.mjs` for `customMessages` and `customUrls` as well.

**Comment by Mouvedia** [comment]
follow-up

We should add a `> [!NOTE]` that states that custom messages, custom URLs and configuration comments are not supported by `checkAgainstRule`.
Or add a new documentation enhancement issue.

**Comment by ybiquitous** [comment]
@Mouvedia You can open an issue for that. 👍🏼
